### DISCUSSION TITLE
Please apply Prettify to all XState Types

### DISCUSSION THREAD

**Discussion by MatthewAry** [statelyai/xstate#4791] [Ideas]
This magical type
```ts
export type Prettify<T> = {
  [K in keyof T]: T[K];
  // eslint-disable-next-line @typescript-eslint/ban-types
} & {};
```
Will take something like this
```ts
types?: MachineTypes<TContext, TEvent, TActor, TAction, TGuard, TDelay, TTag, TInput, TOutput, TEmitted, TTypesMeta> | undefined
```
to be reported as this
```ts
types?: {
    actors?: TActor | undefined;
    actions?: TAction | undefined;
    guards?: TGuard | undefined;
    delays?: TDelay | undefined;
    typegen?: TTypesMeta | undefined;
    context?: TContext | undefined;
    events?: TEvent | undefined;
    children?: undefined;
    tags?: TTag | undefined;
    input?: TInput | undefined;
    output?: TOutput | undefined;
    emitted?: TEmitted | undefined;
} | undefined
```
Most of xstate's types are only good for validation and not any good for documentation. Please use the prettify function so when I am writing code, my IDE can tell me useful information about what can be assigned to different properties in machines and actors.

Thank you!

**Comment by Andarist** [comment]
Tbf, this is somewhat subjective. We are using this type in a couple of places already. I wouldn't use it on `MachineTypes` though. There is a reason why TS itself doesn't just expand the whole content of types in each type display - this would produce unwieldy huge displays.

**Comment by MatthewAry** [comment]
I believe this would greatly improve DX. I'm starting to learn xstate and I am struggling to get any valuable feedback from my IDE because it's not telling me what things can go in what places. Even the generated API docs don't unravel the types in a helpful way either. There is so much prerequisite knowledge needed to truly utilize xstate to it's fullest potential, it's hard for me to imagine a reason why to not do it.

Having the abstracted types unraveled, even if the output is long and unwieldy, would provide more value than not. The TS server often truncates outputs, and when it doesn't, there are tools to make it easier to read.
